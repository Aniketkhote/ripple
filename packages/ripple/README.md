<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://raw.githubusercontent.com/trueadm/ripple/master/assets/ripple-dark.png">
  <img src="https://raw.githubusercontent.com/trueadm/ripple/master/assets/ripple-light.png" alt="Ripple - the elegant UI framework for the web" />
</picture>

# What is Ripple?

> Currently, this project is still in early development, and should not be used in production.

Ripple is a TypeScript UI framework for the web. To find out more, view [Ripple's Github README](https://github.com/trueadm/ripple).
