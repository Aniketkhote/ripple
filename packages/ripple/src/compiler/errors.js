/**
 *
 * @param {string} message
 * @param {string} filename
 * @param {any} node
 */
export function error(message, filename, node) {
	let errorMessage = message;

	if (node && node.loc) {
		// Use GitHub-style range format: filename#L39C24-L39C32
		const startLine = node.loc.start.line;
		const startColumn = node.loc.start.column;
		const endLine = node.loc.end.line;
		const endColumn = node.loc.end.column;

		const rangeInfo = `${filename}#L${startLine}C${startColumn}-L${endLine}C${endColumn}`;
		errorMessage += ` (${rangeInfo})`;
	} else {
		errorMessage += ` (${filename})`;
	}

	throw new Error(errorMessage);
}
