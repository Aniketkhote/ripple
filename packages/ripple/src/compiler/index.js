/** @import { Program } from 'estree' */

import { parse as parse_module } from './phases/1-parse/index.js';
import { analyze } from './phases/2-analyze/index.js';
import { transform_client } from './phases/3-transform/client/index.js';
import { transform_server } from './phases/3-transform/server/index.js';
import { convert_source_map_to_mappings } from './phases/3-transform/segments.js';

/**
 * Parse Ripple source code to ESTree AST
 * @param {string} source
 * @returns {Program}
 */
export function parse(source) {
  return parse_module(source);
}

/**
 * Compile Ripple source code to JS/CSS output
 * @param {string} source
 * @param {string} filename
 * @param {{ mode?: 'client' | 'server' }} [options]
 * @returns {object}
 */
export function compile(source, filename, options = {}) {
  const ast = parse_module(source);
  const analysis = analyze(ast, filename);
  const result = options.mode === 'server'
      ? transform_server(filename, source, analysis)
      : transform_client(filename, source, analysis, false);

  return result;
}

/**
 * Compile Ripple source to Volar mappings for editor integration
 * @param {string} source
 * @param {string} filename
 * @returns {object} Volar mappings object
 */
export function compile_to_volar_mappings(source, filename) {
  // Parse and transform to get the esrap sourcemap
  const ast = parse_module(source);
  const analysis = analyze(ast, filename);
  const transformed = transform_client(filename, source, analysis, true);

  return convert_source_map_to_mappings(transformed.js.map, source, transformed.js.code);
}
