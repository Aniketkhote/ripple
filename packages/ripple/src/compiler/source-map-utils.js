import { decode } from '@jridgewell/sourcemap-codec';

/** @import { PostProcessingChanges, LineOffsets } from './phases/3-transform/client/index.js' */

/** @typedef {{line: number, column: number}} GeneratedPosition */
/** @typedef {Map<string, GeneratedPosition[]>} SourceToGeneratedMap */

/**
 * Build a source-to-generated position lookup map from an esrap source map
 * Applies post-processing adjustments during map building for efficiency
 * @param {object} source_map - The source map object from esrap (v3 format)
 * @param {PostProcessingChanges} post_processing_changes - Optional post-processing changes to apply
 * @param {LineOffsets} line_offsets - Pre-computed line offsets array
 * @returns {SourceToGeneratedMap} Map from "sourceLine:sourceColumn" to array of generated positions
 */
export function build_source_to_generated_map(source_map, post_processing_changes, line_offsets) {
	/** @type {SourceToGeneratedMap} */
	const map = new Map();

	// Decode the VLQ-encoded mappings string
	// @ts-ignore
	const decoded = decode(source_map.mappings);

	/**
	 * Convert line/column position to byte offset
	 * @param {number} line - 1-based line number
	 * @param {number} column - 0-based column number
	 * @returns {number} Byte offset
	 */
	const line_col_to_byte_offset = (line, column) => {
		return line_offsets[line - 1] + column;
	};

	/**
	 * Convert byte offset to line/column
	 * @param {number} offset
	 * @returns {{ line: number, column: number }}
	 */
	const offset_to_line_col = (offset) => {
		// Binary search
		let left = 0;
		let right = line_offsets.length - 1;
		let line = 1;

		while (left <= right) {
			const mid = Math.floor((left + right) / 2);
			if (offset >= line_offsets[mid] && (mid === line_offsets.length - 1 || offset < line_offsets[mid + 1])) {
				line = mid + 1;
				break;
			} else if (offset < line_offsets[mid]) {
				right = mid - 1;
			} else {
				left = mid + 1;
			}
		}

		const column = offset - line_offsets[line - 1];
		return { line, column };
	};

	// decoded is an array of lines, each line is an array of segments
	// Each segment is [generatedColumn, sourceIndex, sourceLine, sourceColumn, nameIndex?]
	for (let generated_line = 0; generated_line < decoded.length; generated_line++) {
		const line = decoded[generated_line];

		for (const segment of line) {
			if (segment.length >= 4) {
				let generated_column = segment[0];
				// just keeping this unused for context
				// const source_index = segment[1]; // which source file (we only have one)
				const source_line = /** @type {number} */ (segment[2]);
				const source_column = /** @type {number} */ (segment[3]);

				// Apply post-processing adjustments if needed
				let adjusted_line = generated_line + 1;
				let adjusted_column = generated_column;

				if (post_processing_changes) {
					const line_change = post_processing_changes.get(adjusted_line);

					if (line_change) {
						// Check if this position is affected by the change
						const pos_offset = line_col_to_byte_offset(adjusted_line, adjusted_column);

						if (pos_offset >= line_change.offset) {
							// Position is on or after the change - apply delta
							const adjusted_offset = pos_offset + line_change.delta;
							const adjusted_pos = offset_to_line_col(adjusted_offset);
							adjusted_line = adjusted_pos.line;
							adjusted_column = adjusted_pos.column;
						}
					}
				}

				// Create key from source position (1-indexed line, 0-indexed column)
				const key = `${source_line + 1}:${source_column}`;

				// Store adjusted generated position
				const gen_pos = { line: adjusted_line, column: adjusted_column };

				if (!map.has(key)) {
					map.set(key, []);
				}
				/** @type {GeneratedPosition[]} */ (map.get(key)).push(gen_pos);
			}
		}
	}

	return map;
}

/**
 * Look up generated position for a given source position
 * @param {number} source_line - 1-based line number in source
 * @param {number} source_column - 0-based column number in source
 * @param {SourceToGeneratedMap} source_to_gen_map - Lookup map
 * @returns {{line: number, column: number} | null} Generated position or null if not found
 */
export function get_generated_position(source_line, source_column, source_to_gen_map) {
	const key = `${source_line}:${source_column}`;
	const positions = source_to_gen_map.get(key);

	if (!positions || positions.length === 0) {
		return null;
	}

	// If multiple generated positions map to same source, return the first
	return positions[0];
}
