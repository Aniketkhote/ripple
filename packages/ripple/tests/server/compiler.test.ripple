import { describe, it, expect } from 'vitest';
import { compile } from 'ripple/compiler'

describe('compiler success tests', () => {


	it('compiles TSInstantiationExpression', () => {
		const source =
`function makeBox<T>(value: T) {
	return { value };
}
const makeStringBox = makeBox<string>;
const stringBox = makeStringBox('abc');
const ErrorMap = Map<string, Error>;
const errorMap = new ErrorMap();`;

		const result = compile(source, 'test.ripple', { mode: 'client' });

		expect(result.js.code).toMatchSnapshot();
	});
});
