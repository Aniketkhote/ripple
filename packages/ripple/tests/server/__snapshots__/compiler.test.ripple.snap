// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler success tests > compiles TSInstantiationExpression 1`] = `
"import * as _$_ from 'ripple/internal/client';

function makeBox(value) {
	return { value };
}

const makeStringBox = (makeBox);
const stringBox = makeStringBox('abc');
const ErrorMap = (Map);
const errorMap = new ErrorMap();"
`;
